Fix non continuous links

jedevc opened the pull request:
The RPM payloadReader attempts to gracefully detect and handle hard links - however, it was relying on the actual data appearing directly after a series of hard links to that data. In practice, however, not all RPMs follow this format, and the actual data may appear far after when it is defined.
    
This patch fixes this behaviour by correctly tracking all inodes seen far, similarly to how the Extract() method already works today.
Additionally, a test is added to ensure this doesn't regress.

This fixes the issue found in #22.

mtharp commented:
Thanks for identifying this issue. Unfortunately I can't merge the selected test RPM due to its license. Ideally I'd like to have a clean spec file for each of the RPMs found in testdata, which not only avoids licensing issues but also means there's an explanation of how it was built. It may be possible to look at the spec file for crypto-policies and figure out how it was built that way. If not, it's fine if this change doesn't have an associated testdata RPM.

The alternate xz library you selected is much, much slower. I deal with some comically large packages from time to time, so going from 15s to almost 2m isn't good. The current library isn't actively maintained, but it's safe enough,, so I would like to keep it for now.

jedevc commented:
I've removed the xz library fix from the patch, definitely out of scope for this PR, woops :+1:

I've removed the test RPM, I'll try and take a look at the spec file and do a follow up patch for a test at some point if I have a spare moment to generate one from scratch.
